File: ABC-336/abc-336/src/bin/d_1.rs
use proconio::input;
use std::cmp::{max, min};

fn main() {
    input! {n:usize, mut a:[usize;n]}
    a.insert(0, 0);

    let mut l = vec![0; n + 2];
    let mut r = vec![0; n + 2];

    for i in 1..=n {
        l[i] = min(a[i], l[i - 1] + 1);
    }

    for i in (0..=n).rev() {
        r[i] = min(a[i], r[i + 1] + 1);
    }
    println!(
        "{}",
        l.iter().zip(r.iter()).map(|(x, y)| x.min(y)).max().unwrap(),
    )
}

File: ABC-189/D_1.py
n = int(input())
S = [input() for _ in range(n)]

dp = [[0] * 2 for _ in range(n + 1)]
# dp[i][0]:iまでにfalseのパターン数
# dp[i][1]:iまでにtrueのパターン数
# dp[i+1][0] = dp[i][0]

dp[0][1] = 1
dp[0][0] = 1

for i in range(n):
    if S[i] == "AND":
        dp[i + 1][0] = dp[i][0] * 2 + dp[i][1]
        dp[i + 1][1] = dp[i][1]
    else:
        dp[i + 1][0] = dp[i][0]
        dp[i + 1][1] = dp[i][0] + dp[i][1] * 2


print(dp[n][1])
